File: lab1-22.02.21/src/main/java/testpassword/lab1/controllers/UserController.java
package testpassword.lab1.controllers;

import lombok.val;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.bind.annotation.*;
import testpassword.lab1.requests.UserReq;
import testpassword.lab1.responses.UserRes;
import testpassword.lab1.security.JWTUtil;
import testpassword.lab1.services.UserService;
import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import java.util.*;

@RestController @RequestMapping(path = "/user")
public class UserController {

    @Autowired private AuthenticationManager auth;
    @Autowired private JWTUtil jwtUtil;
    @Autowired private UserService service;

    @GetMapping(path = "all", produces = "application/json")
    public ResponseEntity<UserRes> getAll() {
        return new ResponseEntity<>(
                UserRes.builder().users(service.getAll()).build(),
                HttpStatus.OK);
    }

    @GetMapping(path = "{userId}", produces = "application/json")
    public ResponseEntity<UserRes> getUser(@PathVariable long userId) {
        return service.get(userId)
                .map(u -> new ResponseEntity<>(
                        UserRes.builder().users(Collections.singletonList(u)).build(),
                        HttpStatus.OK)
                ).orElse(new ResponseEntity<>(
                        UserRes.builder().msg("User with this id didn't exist").build(),
                        HttpStatus.BAD_REQUEST));
    }

    @PostMapping(path = "login", consumes = "application/json", produces = "application/json")
    public ResponseEntity<UserRes> login(@Valid @RequestBody UserReq req) {
        return service.get(req.email)
                .map(u -> {
                    try {
                        auth.authenticate(new UsernamePasswordAuthenticationToken(req.email, req.password));
                        return new ResponseEntity<>(
                                UserRes.builder()
                                        .token(jwtUtil.generateToken(u.getEmail(), Collections.singletonList("USER")))
                                        .build(),
                                HttpStatus.ACCEPTED);
                    } catch (AuthenticationException e) {
                        return new ResponseEntity<>(
                                UserRes.builder().msg("Password incorrect").build(),
                                HttpStatus.BAD_REQUEST);
                    }
                }).orElse(new ResponseEntity<>(
                        UserRes.builder().msg("User didn't exist. Check email and password").build(),
                        HttpStatus.BAD_REQUEST));
    }

    @PutMapping(path = "register", consumes = "application/json", produces = "application/json")
    public ResponseEntity<UserRes> register(@Valid @RequestBody UserReq req) {
        return service.get(req.email)
                .map(u -> new ResponseEntity<>(
                        UserRes.builder().msg("User already exist").build(),
                        HttpStatus.CONFLICT)
                ).orElseGet(() -> {
                    val u = service.add(req.email, req.password, req.name).get();
                    auth.authenticate(new UsernamePasswordAuthenticationToken(req.email, req.password));
                    return new ResponseEntity<>(
                            UserRes.builder()
                                    .token(jwtUtil.generateToken(u.getEmail(), Collections.singletonList("USER")))
                                    .build(),
                            HttpStatus.CREATED);
                });
    }

    @PatchMapping(path = "modify", consumes = "application/json", produces = "application/json")
    public ResponseEntity<UserRes> modify(@Valid @RequestBody UserReq req, HttpServletRequest rawReq) {
        return (service.get(jwtUtil.decode(rawReq)))
                .map(u -> {
                    if (req.email != null)
                        if (service.exist(req.email))
                            return new ResponseEntity<>(
                                    UserRes.builder().msg("Mail is busy").build(),
                                    HttpStatus.CONFLICT);
                        else u.setEmail(req.email);
                    Optional.ofNullable(req.name).ifPresent(u::setName);
                    Optional.ofNullable(req.password).ifPresent(u::setPassword);
                    service.save(u);
                    return new ResponseEntity<>(
                            UserRes.builder().token(jwtUtil.generateToken(req.email, Collections.singletonList("USER"))).build(),
                            HttpStatus.ACCEPTED);
                }).orElse(new ResponseEntity<>(
                        UserRes.builder().msg("User didn't exist").build(),
                        HttpStatus.UNAUTHORIZED));
    }

    @DeleteMapping(path = "delete", consumes = "application/json", produces = "application/json")
    public ResponseEntity<UserRes> delete(@Valid @RequestBody UserReq req, HttpServletRequest rawReq) {
        return (service.get(jwtUtil.decode(rawReq)))
                .map(u -> {
                    try {
                        System.out.println(u.getEmail());
                        auth.authenticate(new UsernamePasswordAuthenticationToken(u.getEmail(), req.password));
                        service.delete(u.getEmail());
                        return new ResponseEntity<>(
                                UserRes.builder().msg("Successful delete account").build(),
                                HttpStatus.OK);
                    } catch (AuthenticationException e) {
                        return new ResponseEntity<>(
                                UserRes.builder().msg("Bad password").build(),
                                HttpStatus.UNAUTHORIZED);
                    }
                }).orElse(new ResponseEntity<>(
                        UserRes.builder().msg("Wrong session token").build(),
                        HttpStatus.UNAUTHORIZED));
    }

    @PostMapping(path = "restore", consumes = "application/json", produces = "application/json")
    public ResponseEntity<UserRes> restore(@Valid @RequestBody UserReq req) {
        return (service.exist(req.email) && service.restorePassword(req.email)) ?
                new ResponseEntity<>(UserRes.builder().msg("Restore email was send").build(), HttpStatus.OK) :
                new ResponseEntity<>(UserRes.builder().msg("Internal error, try later").build(), HttpStatus.BAD_REQUEST);
    }
}

File: lab2-16.03.21/src/main/kotlin/lab2/controllers/UserController.kt
package lab2.controllers

import io.jsonwebtoken.JwtException
import lab2.dtos.UserReq
import lab2.dtos.UserRes
import lab2.models.User
import lab2.security.JWTTokenUtil
import lab2.services.UserService
import org.json.JSONObject
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.LockedException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.AuthenticationException
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpServletRequest
import javax.validation.Valid

@RestController @RequestMapping(path = ["/user"])
class UserController {

    @Autowired private lateinit var auth: AuthenticationManager
    @Autowired private lateinit var jwt: JWTTokenUtil
    @Autowired private lateinit var service: UserService

    private fun ok(block: UserRes.() -> Unit) = ResponseEntity(UserRes().apply(block), HttpStatus.OK)

    private fun bad(block: UserRes.() -> Unit) = ResponseEntity(UserRes().apply(block), HttpStatus.BAD_REQUEST)

    private fun other(status: HttpStatus, block: UserRes.() -> Unit) = ResponseEntity(UserRes().apply(block), status)

    private fun auth(email: String, password: String): String {
        auth.authenticate(UsernamePasswordAuthenticationToken(email, password))
        return jwt.generate(email, listOf("USER"))
    }

    @GetMapping(path = ["all"], produces = ["application/json"])
    fun getAll() =
        ok {
            users = service.getAll()
        }

    @GetMapping(path = ["{userId}"], produces = ["application/json"])
    fun getUser(@PathVariable userId: Long) =
        ok {
            users = listOf(service[userId])
        }

    @GetMapping(path=["spec"], produces = ["application/json"])
    fun getSpecificUsers(@RequestParam ids: String) =
        ok {
            users = ids.split(";").mapNotNull {
                try {
                    service[it.toLong()]
                } catch (e: Exception) {
                    null
                }
            }
        }

    @PostMapping(path = ["extendToken"], produces = ["application/json"])
    fun extendToken(raw: HttpServletRequest) =
        ok {
            if (jwt isExpired (jwt resolve raw)!!)
                token = jwt.generate((service loadUserByUsername (jwt decode raw)).email, listOf("USER"))
            else msg = "You already have actual token"
        }

    @PostMapping(path = ["login"], consumes = ["application/json"], produces = ["application/json"])
    fun login(@Valid @RequestBody req: UserReq) =
        ok {
            service loadUserByUsername req.email
            token = auth(req.email, req.password)
        }

    @PutMapping(path = ["register"], consumes = ["application/json"], produces = ["application/json"])
    fun register(@Valid @RequestBody req: UserReq) =
        try {
            other(HttpStatus.CONFLICT) {
                service loadUserByUsername req.email
                msg = "User ${req.email} already exist"
            }
        } catch (e: EmptyResultDataAccessException) {
            other(HttpStatus.CREATED) {
                service.add(req.email, req.password, req.name).apply { status = User.STATUS.LOCKED }
                token = auth(req.email, req.password)
            }
        }

    @DeleteMapping(path = ["delete"], consumes = ["application/json"], produces = ["application/json"])
    fun delete(@Valid @RequestBody req: UserReq, raw: HttpServletRequest) =
        ok {
            val email = (service loadUserByUsername (jwt decode raw)).email
            auth(email, req.password)
            service delete email
            msg = "Successfully delete account"
        }

    @PutMapping(path = ["modify"], consumes = ["application/json"], produces = ["application/json"])
    fun modify(@Valid @RequestBody req: UserReq, raw: HttpServletRequest) =
        ok {
            val email = (service loadUserByUsername (jwt decode raw)).email
            auth(email, req.password)
            service.modify(email, JSONObject(req.payload).toMap().mapValues { it.toString() })
        }

    @PatchMapping(path = ["reset"], consumes = ["application/json"], produces = ["application/json"])
    fun resetPassword(@Valid @RequestBody req: UserReq) =
        ok {
            msg = if (service resetPassword req.email)
                "Temp password sent to your email" else "Error on server while reseting password, try later"
        }

    @ExceptionHandler(Exception::class)
    fun handleErrors(req: HttpServletRequest, e: Exception) =
        bad {
            msg = when (e) {
                is LockedException -> "Your account locked. Maybe you forgot to change temp password or was banned"
                is JwtException -> "Bad token or didn't presented"
                is AuthenticationException -> "Password incorrect"
                is EmptyResultDataAccessException -> "User didn't exist. Check email and password"
                else -> """
                    Unexpected exception, try later or contact support with this message
                    $e
                    $req
                    """.trimIndent()
            }
        }
}